### PR TITLE
chore(weave): Create a system of defining schema-validated objects

### PR DESCRIPTION
**Pull request by tssweeney**
This PR allows developers to define a single pydantic type which is then used in 3 key ways:
1. These objects can be constructed in Python, published to weave, and "gotten" from weave - resulting in the correct type - not a WeaveObject wrapper thing.
2. Construction of these objects can be done over the network by setting the `set_base_object_class` field on the POST request which will enforce server-side validation of these object payloads.
3. The frontend now has fully typed & validated read/write hooks that allows users to build UI experiences that safely read, create, and edit such objects with type safety.

A top level make command `make synchronize-base-object-schemas` ensures the frontend is up to date!

With all this infrustructure, you simply define your types once, like:

`weave/trace_server/interface/base_object_classes/test_only_example.py`:
```python
from pydantic import BaseModel

from weave.trace_server.interface.base_object_classes import base_object_def


class TestOnlyNestedBaseModel(BaseModel):
    a: int


class TestOnlyNestedBaseObject(base_object_def.BaseObject):
    b: int


class TestOnlyExample(base_object_def.BaseObject):
    primitive: int
    nested_base_model: TestOnlyNestedBaseModel
    nested_base_object: base_object_def.RefStr

__all__ = ["TestOnlyExample", "TestOnlyNestedBaseObject"]
```

And you now have:

1. The ability to publish
```python
weave.publish(TestOnlyExample(...))
```

2. The ability to fetch:
```python
ref = weave.publish(TestOnlyExample(...))
assert isinstance(ref.get(), TestOnlyExample)
```

3. The ability to create via HTTP
```bash
curl -X 'POST' \
  'https://trace.wandb.ai/obj/create' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "obj": {
    "project_id": "user/project",
    "object_id": "my_thing",
    "val": {...},
    "set_base_object_class": "TestOnlyExample" # forces `val` to conform to the schema
  }
```

4. The ability to query via HTTP
```bash
curl -X 'POST' \
  'https://trace.wandb.ai/objs/query' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "project_id": "user/project",
  "filter": {
    "latest_only": true,
    "base_object_classes": ["TestOnlyExample"]
  },
}'
```

5. Typesafe React Hooks (Read-side)
```
const res = useBaseObjectInstances("TestOnlyExample", ...)
// res is typed strictly according to the pydantic schema
```

6. Typesafe React Hooks (Write-side)
```
const createFn = useCreateBaseObjectInstance("TestOnlyExample")
createFn({...})
// createFn is strictly typed and validates client-side before sending to the server!
```

See `dev_docs/BaseObjectClasses.md` for a more complete discussion

### CODE REVIEW

File: dev_docs/BaseObjectClasses.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
+# BaseObjectClasses
```
Comment (tssweeney) [inline dev_docs/BaseObjectClasses.md:1]: TODO: revise this now that implementation is complete

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/wfReactInterface/baseObjectClassQuery.ts [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+import {useGetTraceServerClientContext} from './traceServerClientContext';
+import {
+  TraceObjCreateReq,
+  TraceObjCreateRes,
+  TraceObjQueryReq,
+  TraceObjSchema,
+} from './traceServerClientTypes';
+import {Loadable} from './wfDataModelHooksInterface';
+
+// TODO: This should be generated from the registry!
```
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/pages/wfReactInterface/baseObjectClassQuery.ts:19]: This is an outstanding TODO

File: weave/trace_server/clickhouse_trace_server_batched.py [hunk 3]
```diff
@@ -2059,20 +2065,6 @@ def get_kind(val: Any) -> str:
     return "object"
 
 
-def get_base_object_class(val: Any) -> Optional[str]:
```
Comment (tssweeney) [inline weave/trace_server/clickhouse_trace_server_batched.py:2062]: moved to `weave/trace_server/base_object_class_util.py`

File: weave/trace_server/interface/base_object_classes/base_object_registry.py [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+from weave.trace_server.interface.base_object_classes.base_object_def import BaseObject
+from weave.trace_server.interface.base_object_classes.test_only_example import *
+
+BASE_OBJECT_REGISTRY: Dict[str, Type[BaseObject]] = {
+    "TestOnlyExample": TestOnlyExample,
+    "TestOnlyNestedBaseObject": TestOnlyNestedBaseObject,
+}
+
+
+# TODO: Remove this helper
```
Comment (tssweeney) [inline weave/trace_server/interface/base_object_classes/base_object_registry.py:14]: TODO: Remove this

File: weave/trace_server/sqlite_trace_server.py [hunk 5]
```diff
@@ -1205,20 +1211,6 @@ def get_kind(val: Any) -> str:
     return "object"
 
 
-def get_base_object_class(val: Any) -> Optional[str]:
```
Comment (tssweeney) [inline weave/trace_server/sqlite_trace_server.py:1208]: moved to `weave/trace_server/base_object_class_util.py`

**Review by gtarpenning — APPROVED** [review]
Sweet, excited to see this in action.

File: weave/scripts/generated_base_object_class_schemas.json [hunk 6]
```diff
@@ -0,0 +1,114 @@
+{
```
[the remark this answers is not in the document]
Comment (tssweeney) [reply weave/scripts/generated_base_object_class_schemas.json:1]: moved to `weave/trace_server/interface/base_object_classes/generated/generated_base_object_class_schemas.json`

File: weave/trace_server/base_object_class_util.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
+from typing import Any, Optional, Tuple
+
+from pydantic import BaseModel
+
+from weave.trace_server.interface.base_object_classes.base_object_registry import (
+    BASE_OBJECT_REGISTRY,
+)
+
+base_object_class_names = ["BaseObject", "Object"]
```
[the remark this answers is not in the document]
Comment (tssweeney) [reply weave/trace_server/base_object_class_util.py:9]: added comment

File: weave/trace_server/base_object_class_util.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+                    if val["_bases"][-1] == "BaseModel":
+                        if val["_bases"][-2] in base_object_class_names:
+                            if len(val["_bases"]) > 2:
+                                return val["_bases"][-3]
+                            elif "_class_name" in val:
+                                return val["_class_name"]
+    return None
+
+
+def process_incoming_object(
```
[the remark this answers is not in the document]
Comment (tssweeney) [reply weave/trace_server/base_object_class_util.py:26]: added

File: weave/trace_server/base_object_class_util.py [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+    dump["_bases"] = bases
+    return dump
+
+
+def _general_dump(val: Any) -> Any:
+    if isinstance(val, BaseModel):
+        return dump_base_object(val)
+    elif isinstance(val, dict):
+        return {k: _general_dump(v) for k, v in val.items()}
+    elif isinstance(val, list):
```
[the remark this answers is not in the document]
Comment (tssweeney) [reply weave/trace_server/base_object_class_util.py:87]: added

File: weave/trace_server/interface/base_object_classes/test_only_example.py [hunk 10]
```diff
@@ -0,0 +1,26 @@
+from pydantic import BaseModel
+
+from weave.trace_server.interface.base_object_classes import base_object_def
+
+
+class TestOnlyNestedBaseModel(BaseModel):
```
[the remark this answers is not in the document]
Comment (tssweeney) [reply weave/trace_server/interface/base_object_classes/test_only_example.py:6]: 👍

**Pull request merged**